app/routes.py

```diff
@@ -1,2 +1,219 @@
+
+    return jsonify({"task" : new_task.to_dict()}), 201
+
+def sort_filter_tasks():
+    sort_query = request.args.get("sort")
+    if sort_query == "desc":
+        tasks = Task.query.order_by(Task.title.desc())
+    elif sort_query == "asc":
+        tasks = Task.query.order_by(Task.title.asc())
+    else:
+        tasks = Task.query.all()
+    
+    filter_query = request.args.get("title")
+    if filter_query:
+        tasks = Task.query.filter(Task.title.contains(filter_query))
+    
+    return tasks
+
+@tasks_bp.route("", methods=["GET"])
+def get_tasks():
+    tasks = sort_filter_tasks()
+
+    tasks_response = [task.to_dict() for task in tasks]
+
+    return jsonify(tasks_response), 200
+
+def sort_filter_goals():
+    sort_query = request.args.get("sort")
+    if sort_query == "desc":
+        goals = Goal.query.order_by(Goal.title.desc())
+    elif sort_query == "asc":
+        goals = Goal.query.order_by(Goal.title.asc())
+    else:
+        goals = Goal.query.all()
+
+    filter_query = request.args.get("title")
+    if filter_query:
+        goals = Goal.query.filter(Goal.title.contains(filter_query))
+    
+    return goals
+
+@goals_bp.route("", methods=["GET"])
+def get_goals():
+    goals = sort_filter_goals()
+
+    goal_response = [goal.to_dict() for goal in goals]
+
+    return jsonify(goal_response), 200
+
+@goals_bp.route("/<goal_id>", methods=["GET"])
+@require_goal
+def get(goal):  
+    return jsonify({"goal": goal.to_dict()}), 200
+
+@tasks_bp.route("/<task_id>", methods=["GET"])
+@require_task
+def get(task):  
+    if task.goal_id:
+        return jsonify({"task": task.task_to_dict_w_goal()}), 200
+    return jsonify({"task": task.to_dict()}), 200
+
+@goals_bp.route("/<goal_id>/tasks", methods=["GET"])
+@require_goal
+def get_tasked_goal(goal):
+    return jsonify(goal.goal_task_dict()), 200
+
+@goals_bp.route("/<goal_id>", methods=["PUT"])
+@require_goal
+def put(goal):
+    form_data = request.get_json()
+
+    goal.replace_with_dict(form_data)
+
+    db.session.commit()
+
+    return jsonify({"goal": goal.to_dict()}), 200
+
+@tasks_bp.route("/<task_id>", methods=["PUT"])
+@require_task
+def put(task):
+    form_data = request.get_json()
+
+    task.replace_with_dict(form_data)
+
+    db.session.commit()
+
+    return jsonify({"task": task.to_dict()}), 200
+
+@goals_bp.route("/<goal_id>", methods=["DELETE"])
+@require_goal
+def delete(goal):
+    db.session.delete(goal)
+    db.session.commit()
+    return jsonify ({"details": (f"Goal {goal.id} \"{goal.title}\" successfully deleted")}), 200
+
+@tasks_bp.route("/<task_id>", methods=["DELETE"])
+@require_task
+def delete(task):
+    db.session.delete(task)
+    db.session.commit()
+    return jsonify ({"details": (f"Task {task.id} \"{task.title}\" successfully deleted")}), 200
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+@require_task
+def complete_patch(task):
+    request.get_json()
+
+    task.completed_at = datetime.utcnow()
```

Comment:
I also like using `utcnow()` rather than `now()` for the datetime information. We didn't set up the datetime in the database to store timezone information, so when working with dates and time, it's usually a good idea to pass them around and store them in UTC, and only convert them to local time when needed for display (like in a UI). Working with dates and times is tricky!